src/nemos/basis/_basis.py

```diff
@@ -0,0 +1,1406 @@
+            raise e
+
+    @property
+    def mode(self):
+        """Mode of operation, either ``"conv"`` or ``"eval"``."""
+        return self._mode
+
+    @staticmethod
+    def _apply_identifiability_constraints(X: NDArray):
+        """Apply identifiability constraints to a design matrix `X`.
+
+         Removes columns from `X` until `[1, X]` is full rank to ensure the uniqueness
+         of the GLM (Generalized Linear Model) maximum-likelihood solution. This is particularly
+         crucial for models using bases like BSplines and CyclicBspline, which, due to their
+         construction, sum to 1 and can cause rank deficiency when combined with an intercept.
+
+         For GLMs, this rank deficiency means that different sets of coefficients might yield
+         identical predicted rates and log-likelihood, complicating parameter learning, especially
+         in the absence of regularization.
+
+        Parameters
+        ----------
+        X:
+            The design matrix before applying the identifiability constraints.
+
+        Returns
+        -------
+        :
+            The adjusted design matrix with redundant columns dropped and columns mean-centered.
+        """
+
+        def add_constant(x):
+            return np.hstack((np.ones((x.shape[0], 1)), x))
+
+        rank = np.linalg.matrix_rank(add_constant(X))
+        # mean center
+        X = X - np.nanmean(X, axis=0)
+        while rank < X.shape[1] + 1:
+            # drop a column
+            X = X[:, :-1]
+            # recompute rank
+            rank = np.linalg.matrix_rank(add_constant(X))
+        return X
+
+    @check_transform_input
+    def compute_features(self, *xi: ArrayLike) -> FeatureMatrix:
+        """
+        Apply the basis transformation to the input data.
+
+        This method is designed to be a high-level interface for transforming input
+        data using the basis functions defined by the subclass. Depending on the basis'
+        mode ('eval' or 'conv'), it either evaluates the basis functions at the sample
+        points or performs a convolution operation between the input data and the
+        basis functions.
+
+        Parameters
+        ----------
+        *xi :
+            Input data arrays to be transformed. The shape and content requirements
+            depend on the subclass and mode of operation ('eval' or 'conv').
+
+        Returns
+        -------
+        :
+            Transformed features. In 'eval' mode, it corresponds to the basis functions
+            evaluated at the input samples. In 'conv' mode, it consists of convolved
+            input samples with the basis functions. The output shape varies based on
+            the subclass and mode.
+
+        Notes
+        -----
+        Subclasses should implement how to handle the transformation specific to their
+        basis function types and operation modes.
+        """
+        self._set_num_output_features(*xi)
+        self._set_kernel()
```

Comment:
I would still keep it here for now since any bases needs to have `_set_kernel` and i am going to move this method around in the next PRs. 

the `compute_features` method is basically a `fit_transform`, moving the set_kernel elsewhere means that if you want a different transform type (2D convolution or whatever) you have an extra method in the mixin that you need to add. And you cannot get rid of set_kernel anyways, otherwise it would complicate the logic of the composite classes. 

Overall, I think having it here save as function repeating the same method multiple times. One I will do that, is creating a new method that maps directly to transformer  fit which does both         self._set_num_output_features(*xi) and self._set_kernel(). the compute features will become more readable